hmac/src/lib.rs

```diff
@@ -132,17 +97,35 @@ impl <D> Mac for Hmac<D>
         Self::new_varkey(key.as_slice()).unwrap()
     }
 
+    #[inline]
     fn new_varkey(key: &[u8]) -> Result<Self, InvalidKeyLength> {
+
         let mut hmac = Self {
             digest: Default::default(),
             i_key_pad: Default::default(),
             opad_digest: Default::default(),
-            key: expand_key::<D>(key),
+            key: Default::default(),
         };
-        hmac.i_key_pad = hmac.derive_key(IPAD);
+
+        if key.len() <= hmac.i_key_pad.len() {
+            hmac.i_key_pad[..key.len()].copy_from_slice(key);
+        } else {
+            let mut digest = D::default();
+            digest.process(key);
+            let output = digest.fixed_result();
+            hmac.i_key_pad[..output.len()].copy_from_slice(&output[..output.len()]);
```

Comment:
If output length is bigger than block size this variant will panic. There is no hash functions which have such properties (well, if we exclude XOFs) and this corner case is not covered by the standard, but to make code strictly panic-free we need the old variant. I think that use of `cmp::min` should not influence performance, as `n` will be calculated at compile time and will be equal to `D::OutputSize`.

Probably it's worth to add a small comment to the code explaining this.

Reply:
That makes sense.